node/node_services.go

```diff
@@ -0,0 +1,960 @@
+	vm "github.com/spacemeshos/go-spacemesh/genvm"
+	"github.com/spacemeshos/go-spacemesh/fetch"
+	"github.com/spacemeshos/go-spacemesh/fetch/peers"
+	"github.com/spacemeshos/go-spacemesh/hare3"
+	"github.com/spacemeshos/go-spacemesh/hare3/compat"
+	"github.com/spacemeshos/go-spacemesh/hare3/eligibility"
+	"github.com/spacemeshos/go-spacemesh/hare4"
+	"github.com/spacemeshos/go-spacemesh/layerpatrol"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/malfeasance"
+	"github.com/spacemeshos/go-spacemesh/malfeasance2"
+	"github.com/spacemeshos/go-spacemesh/miner"
+	"github.com/spacemeshos/go-spacemesh/mesh"
+	"github.com/spacemeshos/go-spacemesh/p2p"
+	"github.com/spacemeshos/go-spacemesh/p2p/pubsub"
+	"github.com/spacemeshos/go-spacemesh/proposals"
+	"github.com/spacemeshos/go-spacemesh/proposals/store"
+	"github.com/spacemeshos/go-spacemesh/prune"
+	"github.com/spacemeshos/go-spacemesh/signing"
+	"github.com/spacemeshos/go-spacemesh/syncer"
+	"github.com/spacemeshos/go-spacemesh/syncer/atxsync"
+	"github.com/spacemeshos/go-spacemesh/syncer/blockssync"
+	"github.com/spacemeshos/go-spacemesh/syncer/malsync"
+	"github.com/spacemeshos/go-spacemesh/timesync/peersync"
+	"github.com/spacemeshos/go-spacemesh/tortoise"
+	"github.com/spacemeshos/go-spacemesh/txs"
+)
+
+type initState struct {
+	poetDb *activation.PoetDb
+	postStates *activation.PostStatesWrapper
+	stateDb *vm.VM
+	goldenATXID types.ATXID
+	vrfVerifier signing.VRFVerifier
+	beaconProtocol *beacon.ProtocolDriver
+	trtl *tortoise.Tortoise
+	executor *mesh.Executor
+	mesh *mesh.Mesh
+	pruner *prune.Pruner
+	proposalsStore *store.Store
+	fetcher *fetch.Fetch
+	peerCache *peers.Peers
+	hareOracle *eligibility.Oracle
+	certifier *blocks.Certifier
+	patrol *layerpatrol.LayerPatrol
+	syncer *syncer.Syncer
+	atxHandler *activation.Handler
+	updater *bootstrap.Updater
+	proposalsHandler *proposals.Handler
+	proposalBuilder *miner.ProposalBuilder
+	postSetupMgr *activation.PostSetupManager
+	grpcPostService *grpcserver.PostService
+	poetClients []activation.PoetService
+	nipostBuilder *activation.NIPostBuilder
+	atxBuilder *activation.Builder
+
+	mlog *zap.Logger
+	legacyMalfeasanceLogger *zap.Logger
+	nipostLogger *zap.Logger
+}
+
+type initializerFunc func(ctx context.Context, app *App, state *initState) error
+
+func (app *App) initServices(ctx context.Context) error {
+	var state initState
+	state.peerCache = peers.New()
+	state.patrol = layerpatrol.New()
+
+	for _, initializer := range []initializerFunc{
+		initPoetDb,
+		initPostVerifier,
+		initValidator,
+		initStates,
+		initGoldenATX,
+		initVerifiers,
+		initBeacon,
+		initTortoise,
+		initMesh,
+		initProposalsStore,
+		initFetcher,
+		initHareOracle,
+		initCertifier,
+		initSyncer,
+		initATXHandler,
+		initUpdater,
+		initHare,
+		initProposalsHandler,
+		initBlocksGenerator,
+		initProposalBuilder,
+		initPostService,
+		initPoetClients,
+		initNIPostBuilder,
+		initATXBuilder,
+	} {
```

Comment:
I'm not a fan of this design - this moves compile checks into runtime checks:

If I rearrange the order of this function calls the code will still compile fine but the node will crash on startup because `nils` are passed where pointers to actual types are expected...